src/box/gc.c

```diff
@@ -735,6 +809,238 @@ gc_consumer_name(struct gc_consumer *consumer)
+	const char *object;
+};
+
+/** Mapping from tuple.opts to fields of gc_consumer_def. */
+const struct opt_def gc_consumer_def_opts_reg[] = {
+	OPT_DEF("object", OPT_STRPTR, struct gc_consumer_def, object),
+	OPT_END,
+};
+
+/**
+ * Fill gc_consumer_def with opts from the MsgPack map.
+ * Argument map can be NULL - default options are set in this case.
+ */
+static int
+gc_consumer_def_opts_decode(struct gc_consumer_def *def, const char *map,
+			    struct region *region)
+{
+	def->object = "";
+	if (map == NULL)
+		return 0;
+	return opts_decode(def, gc_consumer_def_opts_reg, &map, region);
+}
+
+/** Build gc_consumer definition from a _gc_consumers' tuple. */
+static struct gc_consumer_def *
+gc_consumer_def_new_from_tuple(struct tuple *tuple, struct region *region)
+{
+	struct gc_consumer_def *def =
+		xregion_alloc_object(region, typeof(*def));
+	memset(def, 0, sizeof(*def));
+	if (tuple_field_uuid(tuple, BOX_GC_CONSUMERS_FIELD_UUID, &def->uuid) != 0)
+		return NULL;
+	if (tt_uuid_is_nil(&def->uuid)) {
+		diag_set(ClientError, ER_INVALID_UUID, tt_uuid_str(&def->uuid));
+		return NULL;
+	}
+	const char *mp_vclock = tuple_field_with_type(
+		tuple, BOX_GC_CONSUMERS_FIELD_VCLOCK, MP_MAP);
+	if (mp_vclock == NULL)
+		return NULL;
+	if (mp_decode_vclock_ignore0(&mp_vclock, &def->vclock) != 0) {
+		diag_set(ClientError, ER_INVALID_VCLOCK);
+		return NULL;
+	}
+	const char *opts = NULL;
+	if (tuple_field(tuple, BOX_GC_CONSUMERS_FIELD_OPTS) != NULL) {
+		opts = tuple_field_with_type(tuple, BOX_GC_CONSUMERS_FIELD_OPTS,
+					     MP_MAP);
+	}
+	if (gc_consumer_def_opts_decode(def, opts, region) != 0)
+		return NULL;
+	return def;
+}
+
+int
+gc_load_consumers(void)
+{
+	if (!gc_has_persistent_consumers())
+		return 0;
+	uint32_t region_svp = region_used(&fiber()->gc);
+	struct space *space = space_by_id(BOX_GC_CONSUMERS_ID);
+	if (space == NULL)
+		return -1;
+	struct index *index = space_index(space, 0);
+	if (index == NULL)
+		return -1;
+	struct iterator *it = index_create_iterator(index, ITER_ALL, NULL, 0);
+	struct tuple *tuple = NULL;
+	int rc = 0;
+	while ((rc = iterator_next(it, &tuple)) == 0 && tuple != NULL) {
+		struct gc_consumer_def *def =
+			gc_consumer_def_new_from_tuple(tuple, &fiber()->gc);
+		if (def == NULL) {
+			say_error("Error while recovering WAL GC consumer");
+			diag_log();
+			continue;
+		}
+		assert(gc_consumer_by_uuid(&def->uuid) == NULL);
+		enum gc_consumer_type type = GC_CONSUMER_UNKNOWN;
+		for (size_t i = 0; i < lengthof(gc_consumer_type_strs); i++) {
+			const char *type_str = gc_consumer_type_strs[i];
+			if (strcmp(def->object, type_str) == 0) {
+				type = (enum gc_consumer_type)i;
+				break;
+			}
+		}
+		struct gc_consumer *consumer =
+			gc_consumer_register_impl(&def->vclock, type,
+						  &def->uuid);
+		if (consumer == NULL)
+			return -1;
+		consumer->is_persistent = true;
+		/* Unref WAL GC delay for each replica having a consumer. */
+		gc_delay_unref();
```

Comment:
I don't think it's correct. We should unref it only for existing replicas. Let's leave delay ref counting as is.

Reply:
Why? This feature is actually workaround that is not needed anymore. If instance is restarted, it doesn't need to delay WAL GC for replicas having persistent consumers. What's about replicas without consumer, we will unref delay when persisting their consumers for the first time.

I'm more concerned that if we leave it as is, persistent WAL GC state won't be actually tested by users until Tarantool 4.0 because `wal_cleanup_delay` will do all the work (I'm pretty sure nobody will switch compat option to `new` without a reason).